### PR TITLE
New testing framework

### LINKED ISSUE #486: Backend testing
**Describe the solution you'd like**
Better backend testing with a well structured testing area 

**Developer story**
As a developer I would like to easily be able to test backend functionallity.

### PR DESCRIPTION
**Pull request by Larspolo**
## Description
I improved our current testing method. Increased by a few percentages, but at least its clearer what's happening

## Summary of changes
- Fixes #486

### CODE REVIEW

File: keep_testing.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (Larspolo) [reply keep_testing.sh]: `bash keep_testing.sh [optional test class]`

File: src/django/test/test_instance.py [hunk 2]
```diff
@@ -1,37 +1,22 @@
...
-        superuser = test.logging_in(self, self.supername, self.superpass)
+class InstanceAPITest(TestCase):
 
-        test.api_get_call(self, '/instance/0/', login=student)
-        test.api_get_call(self, '/instance/0/', login=superuser)
-        self.assertEqual(Instance.objects.all().first().name, 'eJournal')
+    def test_get(self):
+        # Check if PK does not matter
+        resp = api.get(self, 'instance', params={'pk': 0})
+        print(resp)
```
Comment (engelhamer) [inline src/django/test/test_instance.py]: Debug print?

File: src/django/VLE/migrations/0010_auto_20181028_1133.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 2.1.2 on 2018-10-28 11:33
```
Comment (mjvkeulen) [inline src/django/VLE/migrations/0010_auto_20181028_1133.py]: Migrations are a bit messy no?

File: src/django/VLE/views/entry.py [hunk 4]
```diff
@@ -126,86 +111,41 @@ def partial_update(self, request, *args, **kwargs):
...
-        if published is not None:
-            request.user.check_permission('can_publish_grades', assignment)
-            entry.published = published
-            try:
-                entry.save()
-            except ValueError:
-                return response.bad_request('Invalid grade or published state.')
-            if published:
-                Comment.objects.filter(entry=entry).update(published=True)
+            return response.forbidden('The assignment is locked, no entries can changed.')
```
Comment (mjvkeulen) [inline src/django/VLE/views/entry.py]: entries can no longer be edited/changed.

File: src/django/VLE/utils/grading.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import VLE.lti_grade_passback as lti_grade
+from VLE.models import Comment, Entry
+
+
+def publish_all_assignment_grades(assignment, published):
+    """Set published all not None grades from an assignment.
```
Comment (engelhamer) [inline src/django/VLE/utils/grading.py]: Kinda weird sentence. Maybe `publish all grades that are not None for an assignment`?

File: src/django/test/test_assignment.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        api.test_rest(self, 'assignments',
+                      create_params=self.create_params,
+                      create_status=403,
+                      user=factory.Student())
+
+    def test_update(self):
+        assignment = api.create(self, 'assignments', params=self.create_params, user=self.teacher)['assignment']
+
+        # Try to publish the assignment
+        # TODO: Test cannot publish when there are entries inside
```
Comment (engelhamer) [inline src/django/test/test_assignment.py]: Don't you mean _unpublish_?

Comment (engelhamer) [reply src/django/test/test_assignment.py]: There should not be any entries in an unpublished assignment, and a published assignment can not be unpublished once entries have been made.

**Pull request merged**